packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js

```diff
@@ -0,0 +1,573 @@
+  const otherPurpose = soilAmendmentPurposes.find((pu) => pu.key === 'OTHER');
+
+  // Create amendment purpose table
+  await knex.schema.createTable('soil_amendment_task_products_purpose_relationship', (table) => {
+    table
+      .integer('task_products_id')
+      .references('id')
+      .inTable('soil_amendment_task_products')
+      .notNullable();
+    table.integer('purpose_id').references('id').inTable('soil_amendment_purpose').notNullable();
+    table.primary(['task_products_id', 'purpose_id']);
+    table.string('other_purpose');
+    table.check(
+      '(?? IS NOT NULL AND ?? = ??) OR (?? IS NULL)',
+      ['other_purpose', 'purpose_id', otherPurpose.id, 'other_purpose'],
+      'other_purpose_id_check',
+    );
+  });
+
+  // Migrate existing weight data to the new table (reversibly)
+  await knex.raw(`
+    INSERT INTO soil_amendment_task_products (task_id, product_id, weight, weight_unit, application_rate_weight_unit)
+    SELECT task_id, product_id, product_quantity, product_quantity_unit, 'kg/ha'::text FROM soil_amendment_task
+    WHERE product_quantity_unit IN ('${metricWeightUnits.join("', '")}')
+  `);
+
+  await knex.raw(`
+    INSERT INTO soil_amendment_task_products (task_id, product_id, weight, weight_unit, application_rate_weight_unit)
+    SELECT task_id, product_id, product_quantity, product_quantity_unit, 'lb/ac'::text FROM soil_amendment_task
+    WHERE product_quantity_unit IN ('${imperialWeightUnits.join("', '")}')
+  `);
+
+  // Migrate existing volume data to the new table (reversibly)
+  await knex.raw(`
+    INSERT INTO soil_amendment_task_products (task_id, product_id, volume, volume_unit, application_rate_volume_unit)
+    SELECT task_id, product_id, product_quantity, product_quantity_unit, 'l/ha'::text FROM soil_amendment_task
+    WHERE product_quantity_unit IN ('${metricVolumeUnits.join("', '")}')
+  `);
+
+  await knex.raw(`
+    INSERT INTO soil_amendment_task_products (task_id, product_id, volume, volume_unit, application_rate_volume_unit)
+    SELECT task_id, product_id, product_quantity, product_quantity_unit, 'gal/ac'::text FROM soil_amendment_task
+    WHERE product_quantity_unit IN ('${imperialVolumeUnits.join("', '")}')
+  `);
+
+  // Get Product Ids, and Purposes
+  const soilAmendmentTaskPurposes = await knex
+    .select('task_id', 'purpose', 'other_purpose')
+    .table('soil_amendment_task');
+  const soilAmendmentTaskProducts = await knex
+    .select('id', 'task_id')
+    .table('soil_amendment_task_products');
+  const soilAmendmentTaskProductsPurposes = soilAmendmentTaskPurposes.map((task) => {
+    const soilAmendmentTaskProduct = soilAmendmentTaskProducts.find(
+      (pr) => pr.task_id === task.task_id,
+    );
+    const soilAmendmentPurpose = soilAmendmentPurposes.find(
+      (pu) => pu.key === task.purpose.toUpperCase(),
+    );
+    return {
+      task_products_id: soilAmendmentTaskProduct.id,
```

Comment:
The migration fails for me. Do you have any idea what's happening off the top of your head?
![Screenshot 2024-06-06 at 1 49 23 PM](https://github.com/LiteFarmOrg/LiteFarm/assets/33141219/fc507281-4a0f-4df9-a3bf-cf3da2d8dc8d)

Reply:
I think these lines I forgot to add "|| null" appended to the end. Do you have any soil amendment tasks without product_id or purpose? More likely purpose because not required on frontend

Reply:
Oh I have a task without `product_id`! I might have been testing APIs, thank you!!

Reply:
@Duncan-Brain 

> More likely purpose because not required on frontend.

It is required

Reply:
Oh true! Must be somewhat recent update since there are >300 occurrences of null on prod vs product_id ~120 on prod: lite-farm.atlassian.net/browse/LF-4246

Reply:
Sorry about this one yah, I knew about it and totally forgot to test it.